### PR TITLE
fix: Change Redis.go to support explicitely connecting to a Redis Cluster with a single address for GCP compatibility

### LINKED ISSUE #449: Open Saves cannot connect with GCP MemoryStore - Redis Cluster
**What happened**:
When trying to use OpenSaves to connect with GCP MemoryStore - Redis Cluster, as the Redis Cluster is using a single discovery address the Go-Redis UniversalClient is failing to create a ClusterClient and instead is creating a normal Client.

**What you expected to happen**:
I should be able to connect to GCP MemoryStore - Redis Cluster using a single discovery address.

**How to reproduce it (as minimally and precisely as possible)**:
Create a minimal GCP MemoryStore - Redis Cluster and try to connect to it using the discovery address.

**Anything else we need to know?**:

**Environment**:
- Open Saves version: v0.3.0-beta0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration: GCP
- Install method (yaml/helm):
- Troubleshooting guide log(s):
- Others:

### PR DESCRIPTION
**Pull request by Zurvarian**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind feat
> /kind build
> /kind chore
> /kind ci
> /kind docs
> /kind style
> /kind refactor
> /kind perf
> /kind test

/kind fix

**What this PR does / Why we need it**:
When testing Redis connection with GCP managed Redis Cluster we've seen that GCP expects clients to connect using a single discovery IP instead of the set of IPs that is expected by Go-Redis UniversalClient.

**Which issue(s) this PR fixes**:

Closes #449

**Special notes for your reviewer**:

### CODE REVIEW

File: internal/pkg/cache/redis/redis.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ type Redis struct {
 	c redis.UniversalClient
 }
 
+const (
+	redisClusterPrefix = "cluster/"
```
Comment (lorangf) [inline internal/pkg/cache/redis/redis.go:33]: I do not like the implicit prefix hidden in the connection string.

I prefer if you provide two explicit New functions, one for creating a redis server and another for creating a redis cluster.

File: internal/pkg/cache/redis/redis.go [hunk 2]
```diff
@@ -40,16 +45,30 @@ func NewRedis(address string) *Redis {
...
 func NewRedisWithConfig(cfg *config.RedisConfig) *Redis {
-	o := &redis.UniversalOptions{
-		Addrs:           parseRedisAddress(cfg.Address),
-		MinIdleConns:    cfg.MinIdleConns,
-		PoolSize:        cfg.PoolSize,
-		ConnMaxIdleTime: cfg.IdleTimeout,
-		ConnMaxLifetime: cfg.MaxConnAge,
+	var c redis.UniversalClient
+
+	if isRedisClusterAddress(cfg.Address) {
```
Comment (lorangf) [inline internal/pkg/cache/redis/redis.go:50]: Please use a different config variable to indicate to the service whether or not the redis server is in cluster mode or not.

**Review by lorangf — CHANGES_REQUESTED** [review]
Change the way we determine if the redis server is in cluster mode on not based on an additional config variable.

File: internal/pkg/cache/redis/redis.go [hunk 3]
```diff
@@ -33,23 +43,40 @@ type Redis struct {
...
-		Addrs:           parseRedisAddress(cfg.Address),
-		MinIdleConns:    cfg.MinIdleConns,
-		PoolSize:        cfg.PoolSize,
-		ConnMaxIdleTime: cfg.IdleTimeout,
-		ConnMaxLifetime: cfg.MaxConnAge,
+	var c redis.UniversalClient
+
+	if cfg.RedisMode == RedisModeCluster {
+		o := &redis.ClusterOptions{
+			Addrs:           parseRedisClusterMultiAddress(cfg.Address),
```
Comment (lorangf) [inline internal/pkg/cache/redis/redis.go:58]: How is this working with the Redis Cluster discovery endpoint? Is the discovery endpoint a list of IPs, or is it always one IP?

**Review by lorangf — COMMENTED** [review]
Looks good but I have a question regarding the discovery endpoint.

File: internal/pkg/cache/redis/redis.go (hunk 3 above)
Comment (Zurvarian) [reply internal/pkg/cache/redis/redis.go:58]: https://cloud.google.com/memorystore/docs/cluster/cluster-node-specification#discovery_endpoint

TL;DR: The discovery endpoint guarantees a consistent IP for the entire lifecycle of the Redis Cluster, independently of the nodes going up or down. Once the client connects to the endpoint, the client starts receiving packets with the master and replica nodes it must create a connection with.

File: internal/pkg/cache/redis/redis.go [hunk 4]
```diff
@@ -65,11 +95,11 @@ func NewRedisWithConfig(cfg *config.RedisConfig) *Redis {
 	}
 }
 
-// Parse the input Redis address by splitting the list of addresses separated by commas (,)
-func parseRedisAddress(address string) []string {
+// parseRedisClusterMultiAddress Parse the input Redis address by splitting the list of addresses separated by commas (,)
```
Comment (lorangf) [inline internal/pkg/cache/redis/redis.go:98]: Can we change this back to the original function name if this also works for both redis server and redis cluster?

**Review by lorangf — CHANGES_REQUESTED** [review]
Small tweak.

File: internal/pkg/cache/redis/redis.go (hunk 4 above)
Comment (Zurvarian) [reply internal/pkg/cache/redis/redis.go:98]: Sure thing
